lib/Frontend/CompilerInvocation.cpp

```diff
@@ -166,36 +134,61 @@ class ArgsToFrontendInputsConverter {
   bool convert() {
     if (enforceFilelistExclusion())
       return true;
-    getFilesFromCommandLine();
-    if (getFilesFromFilelist())
+    bool hadError = getFilesFromCommandLine();
+    if (hadError)
+      return true;
+    hadError = getFilesFromFilelist();
+    if (hadError)
       return true;
-    return setPrimaryFiles();
+
+    llvm::StringSet<> PrimaryFiles = getPrimaries();
+
+    for (auto file : Files) {
+      bool isPrimary = PrimaryFiles.count(file) > 0;
+      Inputs.addInput(InputFile(file, isPrimary));
+      if (isPrimary)
+        PrimaryFiles.erase(file);
+    }
+    for (auto &file : PrimaryFiles) {
+      // Catch "swiftc -frontend -c -filelist foo -primary-file
+      // some-file-not-in-foo".
+      assert(doesCommandLineIncludeFilelist() &&
+             "Missing primary with no filelist");
+      Diags.diagnose(SourceLoc(), diag::error_primary_file_not_found,
+                     file.getKey(), filelistPath());
+    }
+    return !PrimaryFiles.empty();
   }
 
 private:
   bool enforceFilelistExclusion() {
-    if (Args.hasArg(options::OPT_INPUT) && FilelistPathArg != nullptr) {
+    if (Args.hasArg(options::OPT_INPUT) && doesCommandLineIncludeFilelist()) {
       Diags.diagnose(SourceLoc(),
                      diag::error_cannot_have_input_files_with_file_list);
       return true;
     }
     return false;
   }
-  void getFilesFromCommandLine() {
+
+  bool getFilesFromCommandLine() {
+    bool hadDuplicates = false;
     for (const Arg *A :
          Args.filtered(options::OPT_INPUT, options::OPT_primary_file)) {
-      StringRef file = A->getValue();
-      if (A->getOption().matches(options::OPT_INPUT))
-        addFile(file, Whence::SecondaryFromCommandLine);
-      else if (A->getOption().matches(options::OPT_primary_file))
-        addFile(file, Whence::PrimaryFromCommandLine);
-      else
-        llvm_unreachable("Unknown input-related argument!");
+      if (A->getOption().matches(options::OPT_primary_file) &&
+          mustPrimaryFilesOnCommandLineAlsoAppearInFileList())
+        continue;
+      hadDuplicates = addFile(A->getValue()) || hadDuplicates;
     }
+    return false; // Don't bail out for duplicates, too many tests depend on it.
```

Comment:
Nitpick: can this be marked FIXME, just to make it clear that this is not an ideal state?

Reply:
Done!